Mathlib/Tactic/Congrm.lean

```diff
@@ -0,0 +1,137 @@
+/-
+Copyright (c) 2023 Moritz Doll, Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Moritz Doll, Gabriel Ebner, Damiano Testa
+-/
+
+import Mathlib.Tactic.Basic
+import Mathlib.Tactic.Congr!
+import Mathlib.Logic.Basic
+
+/-!
+# The `congrm` tactic
+
+to be written
+ -/
+
+namespace Mathlib.Tactic
+open Lean Parser Tactic Elab Tactic Meta
+
+initialize registerTraceClass `congrm
+
+/--
+Assume that the goal is of the form `lhs = rhs` or `lhs ↔ rhs`.
+`congrm e` takes an expression `e` containing placeholders `_` and scans `e, lhs, rhs` in parallel.
+
+It matches both `lhs` and `rhs` to the pattern `e`, and produces one goal for each placeholder,
+stating that the corresponding subexpressions in `lhs` and `rhs` are equal.
+
+Examples:
+```lean
+example {a b c d : ℕ} :
+  Nat.pred a.succ * (d + (c + a.pred)) = Nat.pred b.succ * (b + (c + d.pred)) := by
+  congrm Nat.pred (Nat.succ _) * (_ + _)
+/-  Goals left:
+⊢ a = b
+⊢ d = b
+⊢ c + a.pred = c + d.pred
+-/
+  sorry
+  sorry
+  sorry
+
+example {a b : ℕ} (h : a = b) : (λ y : ℕ => ∀ z, a + a = z) = (λ x => ∀ z, b + a = z) := by
+  congrm λ x => ∀ w, _ + a = w
+  -- produces one goal for the underscore: ⊢ a = b
+  exact h
+```
+-/
+syntax (name := congrM) "congrm " term : tactic
+
+section util
+
+/-- Given a function, extract the explicit arguments. -/
+def _root_.Lean.Expr.getExplicitArgs (f : Expr) : MetaM (Array Expr) := do
+  let args := f.getAppArgs
+  let pinfo := (← Lean.Meta.getFunInfo f.getAppFn').paramInfo
+  return (pinfo.zip args).filterMap (λ arg => if arg.1.isExplicit then some arg.2 else none)
+
+/-- try to close the goal using refl, return the `mvarId` if it fails-/
+def tryRefl (goal : MVarId) : MetaM (Option MVarId) := do
+  let res ← observing? do
+    goal.refl
+  if res == none then return some goal else return none
+
+/-- to be moved somewhere else-/
+def _root_.Lean.MVarId.applyWithFreshMVarLevels (goal : MVarId) (lem : Name) :
+    MetaM (List MVarId) := do
+  goal.apply (← mkConstWithFreshMVarLevels lem)
+
+end util
+
+private partial def telescopingFn (lem : Name) (goal : MVarId) (x : Expr) : MetaM MVarId := do
+  let userName ← x.fvarId!.getUserName
```

Comment:
If you're immediately calling `.fvarId!`, wouldn't it make more sense to have an `FVarId` argument?